File: src/main/java/org/hisp/dhis/rules/models/RuleVariablePreviousEvent.java
package org.hisp.dhis.rules.models;

import com.google.auto.value.AutoValue;
import org.hisp.dhis.rules.Option;
import org.hisp.dhis.rules.RuleVariableValue;
import org.hisp.dhis.rules.RuleVariableValueMapBuilder;
import org.hisp.dhis.rules.Utils;

import javax.annotation.Nonnull;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.hisp.dhis.rules.Utils.getLastUpdateDateForPrevious;

@AutoValue
public abstract class RuleVariablePreviousEvent
    extends RuleVariableDataElement
{

    @Nonnull
    public static RuleVariablePreviousEvent create(@Nonnull String name,
       @Nonnull String dataElement, @Nonnull RuleValueType valueType, boolean useCodeForOptionSet, List<Option> options)
    {
        return new AutoValue_RuleVariablePreviousEvent( name, useCodeForOptionSet, options , dataElement, valueType );
    }

    @Override
    public Map<String, RuleVariableValue> createValues( RuleVariableValueMapBuilder builder,
        Map<String, List<RuleDataValueHistory>> allEventValues,
        Map<String, RuleAttributeValue> currentEnrollmentValues,
        Map<String, RuleDataValue> currentEventValues )
    {
        Map<String, RuleVariableValue> valueMap = new HashMap();

        RuleVariableValue variableValue = null;
        List<RuleDataValueHistory> ruleDataValues = allEventValues.get( this.dataElement() );
        if ( builder.ruleEvent != null && ruleDataValues != null && !ruleDataValues.isEmpty() )
        {
            for ( RuleDataValueHistory ruleDataValue : ruleDataValues )
            {
                // We found preceding value to the current currentEventValues,
                // which is assumed to be best candidate.
                if ( builder.ruleEvent.eventDate().compareTo( ruleDataValue.getEventDate() ) > 0
                    || (
                        builder.ruleEvent.eventDate().compareTo( ruleDataValue.getEventDate() ) == 0 &&
                        builder.ruleEvent.createdDate().compareTo( ruleDataValue.getCreatedDate() ) > 0
                        ))
                {
                    String optionValue = this.useCodeForOptionSet() ? ruleDataValue.getValue() : getOptionName( ruleDataValue.getValue() );

                    variableValue = RuleVariableValue.create( optionValue, this.dataElementType(),
                        Utils.valuesForHistory( ruleDataValues ),
                        getLastUpdateDateForPrevious( ruleDataValues, builder.ruleEvent ) );
                    break;
                }
            }
        }

        if ( variableValue == null )
        {
            variableValue = RuleVariableValue.create( this.dataElementType() );
        }

        valueMap.put( this.name(), variableValue );
        return valueMap;
    }
}

File: src/jvmMain/kotlin/org/hisp/dhis/rules/models/RuleVariablePreviousEvent.kt
package org.hisp.dhis.rules.models

import org.hisp.dhis.rules.*

class RuleVariablePreviousEvent(
    val name: String,
    val useCodeForOptionSet: Boolean,
    val options: List<Option>,
    val dataElement: String,
    val dataElementType: RuleValueType
) : RuleVariableDataElement {
    override fun dataElement(): String {
        return dataElement
    }

    override fun dataElementType(): RuleValueType {
        return dataElementType
    }

    override fun options(): List<Option> {
        return options
    }

    override fun createValues(
        builder: RuleVariableValueMapBuilder,
        allEventValues: Map<String, List<RuleDataValue>>,
        currentEnrollmentValues: Map<String, RuleAttributeValue>,
        currentEventValues: Map<String, RuleDataValue>
    ): Map<String, RuleVariableValue> {
        val valueMap: MutableMap<String, RuleVariableValue> = HashMap()
        var variableValue: RuleVariableValue? = null
        val ruleDataValues = allEventValues[dataElement]
        if (builder.ruleEvent != null && ruleDataValues != null && !ruleDataValues.isEmpty()) {
            for (ruleDataValue in ruleDataValues) {
                // We found preceding value to the current currentEventValues,
                // which is assumed to be best candidate.
                if (builder.ruleEvent!!.eventDate().compareTo(ruleDataValue.eventDate) > 0) {
                    val optionValue =
                        if (useCodeForOptionSet) ruleDataValue.value else getOptionName(ruleDataValue.value)!!
                    variableValue = RuleVariableValue(
                        dataElementType, optionValue,
                        values(ruleDataValues),
                        getLastUpdateDateForPrevious(ruleDataValues, builder.ruleEvent!!)
                    )
                    break
                }
            }
        }
        if (variableValue == null) {
            variableValue = RuleVariableValue(dataElementType)
        }
        valueMap[name] = variableValue
        return valueMap
    }
}
